rfcs/002-merge-types-terms-syntax.md

```diff
@@ -0,0 +1,452 @@
+
+- **Where to allow a tail**: the only part that is in `Type` but not in `Term`
+  is precisely the record tail. Should we allow tails in records that aren't
+  the embedding of a type, such as `{foo = fun x => x; a}`? More in the spirit
+  of actually mixing terms and types, we could indeed allow it, having records
+  storing their tail in a specific attribute, and adapt contract application
+  to take it into account. But this implies giving semantics to things like
+  `record.is_empty { ; a}` or `{ ; a} & { ; b}`. Those are actually
+  interesting questions outside of the context of this RFC (see
+  [201](https://github.com/tweag/nickel/issues/201)).
+
+  While allowing tails in arbitrary record may look more uniform, it requires
+  more design thinking. There isn't any obvious practical usage that seems to
+  require it today. In practice we expect a record with a tail to actually be a
+  record type and be used as a type annotation or applied as a contract derived
+  from a type, but nothing else. Finally, we can always allow it later on, as it
+  should (or, rather, must) be backward-compatible with the simple solution.
+
+  In consequence, we choose to only allow tail on a record that is actually a
+  type in the current syntax. In fact we can allow for slightly more room here
+  later, so we abstract this notion in the predicate `is_record_type : UniTerm
+  -> Bool` and in the end, **propose to set the syntax to require `r = {decl1,
+  ..., decln; tail} => is_record_type({decl1, ..., decln})`**. In this RFC,
+  we'll define `is_record_type(t)` in the obvious way such as if `t` is a record
+  type in the current syntax, i.e. `is_record_type(t) <=> exists. T : Type, T
+  record type, t = !T`. In particular `{foo = fun x => x; a}` is now rejected.
+
+The `{_: Num}` syntax is not ambiguous and can be special cased, as it is
+already in the current `Type` syntax (it only requires a one character
+look-ahead to distinguish).
+
+#### Enums
+
+With the first release in mind, we propose to just disable support for enum
+altogether. This is a handy but hardly fundamental feature, and this lets us
+more time to find a good syntax replacement. We can also disable `switch`
+temporarily, as currently its only usage is for enums.
+
+### Translation
+
+Let's now define the `term` and `type` translations. Our design principle is to
+try to have them just be simple syntactic translations that insert the implicit
+`#` and `§` appropriately, without having to recurse into terms or do anything
+complex, whenever possible.
+
+The two functions must satisfy the following coherence laws with `! : Term +
+Type -> UniTerm` the embedding of the current syntax to the new unified syntax
+(erasing `#` and `§`, and changing `|` for tails to `;`) and `~` some flavour of
+operational equivalence (we won't prove those rules formally, as this a mere
+design guideline, so we also restrain from giving a formal definition for `~`).
```

Comment:
I didn't understand the part of the paragraph starting at “`~` some flavour of”.

Reply:
It's a convoluted way of saying: I don't want to use term equality as it is too strict, but some kind of operational equivalence instead. However I don't want to define it formally. It wouldn't really make sense, given there is no formal specification of Nickel at this point anyway. Maybe I can just drop what's inside the parentheses.

Reply:
Maybe cut the sentence in smaller pieces as well. Now that you've told me what to look for, I managed to parse it (after two or three failed attempts). It's perfectly grammatical, but I got quite lost along the way.